casper/src/main/scala/coop/rchain/casper/util/comm/FairRoundRobinDispatcher.scala

```diff
@@ -28,11 +27,13 @@ class FairRoundRobinDispatcher[F[_]: Sync: Log, S: Show, M: Show: Eq](
   def dispatch(source: S, message: M): F[Unit] =
     filter(message).flatMap {
       case Handle =>
-        ensureSourceExists(source) >>
-          isDuplicate(source, message).ifM(
-            Log[F].info(s"Dropped duplicate message ${message.show} from ${source.show}"),
-            enqueueMessage(source, message) >> handleMessages
-          )
+        lock.withPermit(
+          ensureSourceExists(source) >>
+            isDuplicate(source, message).ifM(
+              Log[F].info(s"Dropped duplicate message ${message.show} from ${source.show}"),
+              enqueueMessage(source, message) >> handleMessages
```

Comment:
Am I missing something or `handleMessages` will still process all messages in the queue before lock is released?

Reply:
we have a filter here https://github.com/rchain/rchain/pull/2838/files#diff-ac0666f5850632bde5717addd98fa3f7R28.
So now only handling `blockHashMessage` still is limited by the lock

Reply:
I'm not sure that I see how filter helps. Is it call to `handleMessages` process all messages in the queue? If so, the lock will not be released until `handleMessages` returns so any new `Handle` that passes filter will block.

Reply:
Only `blockHashMessage` is put into the queue. Other messages like `BlockMessage`, `BlockRequest`, `HasBlockRequest`, `ForkChoiceTipRequest` and `HasBlock` are filtered to be case `Pass`(https://github.com/rchain/rchain/pull/2838/files#diff-ac0666f5850632bde5717addd98fa3f7L36) and handle.

That's my understanding of how `FairRoundRobinDispatcher` works. Does that make sense to you?

Reply:
I see now what you mean, `BlockHashMessage` is now only message locked when it's processed. What I'm not sure is do we need to wait for `handleMessages` to finish to release the lock.
What you did already greatly reduces waiting on synchronization.